File: Java/src/main/java/at/tugraz/experiments/DataLoadToMemorySingle.java
package at.tugraz.experiments;

import at.tugraz.runtime.ObjectReader;
import at.tugraz.util.Const;
import at.tugraz.util.RootData;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class DataLoadToMemorySingle {

    public static void main(String[] args) {

        String inDataPath = System.getProperty("inDataPath");
        String seqRand = System.getProperty("seqRand");
        int nrow = Integer.parseInt(System.getProperty("nrow"));
        ObjectReader reader = new ObjectReader(inDataPath, "Kryo");
        int size = Const.BATCHSIZE;

        if (seqRand.equalsIgnoreCase("sequential")) {
            for (int i = 0; i < nrow; ) {
                RootData[] rd = reader.readObjects(i, size);
                i += rd.length;
                size = Math.min(nrow - i + 1, Const.BATCHSIZE);
            }
        }
        else{
            String randomDataPath = System.getProperty("randomDataPath");
            int[] randomIDs= new int[nrow];
            try (BufferedReader br = new BufferedReader(new FileReader(randomDataPath))) {
                String line;
                int index=0;
                while ((line = br.readLine()) != null) {
                    randomIDs[index++] = Integer.parseInt(line);
                }
                for (int i: randomIDs ) {
                    reader.readObjects(i, 1);
                }

            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}

File: Java/src/main/cpp/experiments/DataLoadToMemorySingle.cpp
#include <iostream>
#include "ObjectReader.h"
#include "Const.h"
#include "DataReader.h"

using namespace std;

int main(int argc, char *argv[]) {

    string inDataPath = argv[1];
    string seqRand = argv[2];
    int nrow = atoi(argv[3]);

    ObjectReader *reader = new ObjectReader(inDataPath, "HandCoded");
    int size = BATCHSIZE;
    if (strcasecmp(seqRand.c_str(), "sequential") == 0) {
        for (int i = 0; i < nrow;) {
            TweetStatus **tweets = new TweetStatus *[size];
            int rdSize = reader->readObjects(i, size, tweets);
            for (int j = 0; j < rdSize; j++) {
                delete tweets[j];
            }
            i += rdSize;
            size = min(nrow - i, BATCHSIZE);
            delete[] tweets;
        }
    }
//    else {
//        string randomDataPath = argv[4];
//        ifstream infile;
//        infile.open(randomDataPath);
//        if (!infile.is_open()) {
//            throw runtime_error("cannot find random list file!!.");
//        }
//
//        string line;
//        int randomIDs[nrow];
//        int index = 0;
//        while (getline(infile, line)) {
//            randomIDs[index++] = atol(line.c_str());
//        }
//
//        switch (reader->getMethod()) {
//            case HANDCODED:
//            case BOOST:
//            case BSON: {
//                TweetStatus **tweets = new TweetStatus *[reader->getRlen()];
//                int index = 0;
//                for (int i: randomIDs) {
//                    tweets[index] = reader->readObject(i);
//                    delete tweets[index++];
//                }
//                delete[] tweets;
//                break;
//            }
//            case INPLACE: {
//                TweetStatusIP **tweets = new TweetStatusIP *[reader->getRlen()];
//                int index = 0;
//                for (int i: randomIDs) {
//                    tweets[index] = reader->readObjectIP(i);
//                    delete tweets[index++];
//                }
//                delete[] tweets;
//                break;
//            }
//            case PROTOBUF: {
//                TweetStatusProto **tweets = new TweetStatusProto *[reader->getRlen()];
//                int index = 0;
//                for (int i: randomIDs) {
//                    tweets[index] = reader->readObjectProto(i);
//                    delete tweets[index++];
//                }
//                delete[] tweets;
//                break;
//            }
//            case FLATBUF: {
//                TweetStatusFlatBuffers **tweets = new TweetStatusFlatBuffers *[reader->getRlen()];
//                int index = 0;
//                for (int i: randomIDs) {
//                    tweets[index] = reader->readObjectFlatBuffers(i);
//                    delete tweets[index++];
//                }
//                delete[] tweets;
//                break;
//            }
//        }
//    }

    delete reader;
    return 0;
}
